hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/clustering/HoodieFlinkClusteringJob.java

```diff
@@ -335,5 +391,17 @@ public void shutdownAsyncService(boolean error) {
     public void shutDown() {
       shutdownAsyncService(false);
     }
+
+    /**
+     * Execute a dummy pipeline to prevent "no execute() calls" exceptions from being thrown if
+     * clustering is not performed.
+     */
```

Comment:
What do you mean for `no execute() calls` then ?

Reply:
<img width="1406" alt="image" src="https://user-images.githubusercontent.com/6312314/188038599-9df06074-0ad3-488b-a07b-5076476a3458.png">

Reply:
Should we just stop submit the job to cluster then ?

Reply:
How do we do that? 

Do we call `shutDown(false)` to achieve this?

Reply:
How do you submit the job, it weird what the no `execute()` exception throws because we return early in the executor worker thread.

Reply:
Hmmm, CMIIW, `execute()` is only triggered at the end of the `cluster()` function.

```java
env.execute("flink_hudi_clustering_" + clusteringInstant.getTimestamp());
```

If `cluster()` function terminates before reaching the end, the `execute()` function will not be called.

Hence, there will be a `no execute() exception` as shown in the image. 

Jobs are submitted via Flink web portal using the hoodie flink bundle jar.

Reply:
Seems it is related with the way to submit the job, does the same error happens for CLI submission(for example, the service mode), if it is not, i would suggest to remove the dummy pipeline to avoid confusion.

Reply:
I have not tested it on service mode before since our use case doesn't call for it. 

Is there a better way around this?

Reply:
@danny0405 Error message will not happen in service mode. 

When running under service mode, the job will wait for a plan to be created.

Reply:
While I have yet to test it on service mode, from my understanding of the code:

Given that an error is already thrown here (non service mode):
```java
try {
        compactionScheduleService.compact();
      } catch (Exception e) {
        LOG.error("Got error running delta sync once. Shutting down", e);
        throw e;
      } finally {
        LOG.info("Shut down hoodie flink compactor");
      }
```

The error will be thrown as a `HoodieException` in the same service mode execution block.

```java
try {
  compact();
  Thread.sleep(cfg.minCompactionIntervalSeconds * 1000);
} catch (Exception e) {
  LOG.error("Shutting down compaction service due to exception", e);
  error = true;
  throw new HoodieException(e.getMessage(), e);
}
```

The error logs in text form in non-service mode execution:
```java
2022-09-06 11:59:45,633 INFO  org.apache.hudi.common.table.HoodieTableMetaClient           [] - Loading HoodieTableMetaClient from hdfs://hudi_mor_compaction_test/
2022-09-06 11:59:45,638 INFO  org.apache.hudi.common.table.HoodieTableConfig               [] - Loading table properties from hdfs://hudi_mor_compaction_test/.hoodie/hoodie.properties
2022-09-06 11:59:45,643 INFO  org.apache.hudi.common.table.HoodieTableMetaClient           [] - Finished Loading Table of type MERGE_ON_READ(version=1, baseFileFormat=PARQUET) from hdfs://hudi_mor_compaction_test/
2022-09-06 11:59:45,643 INFO  org.apache.hudi.common.table.HoodieTableMetaClient           [] - Loading Active commit timeline for hdfs://hudi_mor_compaction_test/
2022-09-06 11:59:46,012 INFO  org.apache.hudi.common.table.timeline.HoodieActiveTimeline   [] - Loaded instants upto : Option{val=[==>20220906115739354__deltacommit__INFLIGHT]}
2022-09-06 11:59:46,018 INFO  org.apache.hudi.common.table.view.FileSystemViewManager      [] - Creating View Manager with storage type :REMOTE_FIRST
2022-09-06 11:59:46,018 INFO  org.apache.hudi.common.table.view.FileSystemViewManager      [] - Creating remote first table view
2022-09-06 11:59:46,020 INFO  org.apache.hudi.sink.compact.HoodieFlinkCompactor            [] - Hoodie Flink Compactor running only single round
2022-09-06 11:59:46,361 INFO  org.apache.hudi.common.table.timeline.HoodieActiveTimeline   [] - Loaded instants upto : Option{val=[==>20220906115739354__deltacommit__INFLIGHT]}
2022-09-06 11:59:46,363 INFO  org.apache.hudi.client.RunsTableService                      [] - No compaction plan scheduled, turns on the compaction plan schedule with --schedule option
2022-09-06 11:59:46,363 INFO  org.apache.hudi.sink.compact.HoodieFlinkCompactor            [] - Shut down hoodie flink compactor
2022-09-06 11:59:46,420 ERROR org.apache.flink.client.deployment.application.ApplicationDispatcherBootstrap [] - Application failed unexpectedly. Killing cluster...
java.util.concurrent.CompletionException: org.apache.flink.client.deployment.application.ApplicationExecutionException: The application contains no execute() calls.
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292) ~[?:1.8.0_232]
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308) ~[?:1.8.0_232]
	at java.util.concurrent.CompletableFuture.uniCompose(CompletableFuture.java:957) ~[?:1.8.0_232]
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:940) ~[?:1.8.0_232]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488) ~[?:1.8.0_232]
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1990) ~[?:1.8.0_232]
	at org.apache.flink.client.deployment.application.ApplicationDispatcherBootstrap.runApplicationEntryPoint(ApplicationDispatcherBootstrap.java:263) ~[flink-dist_2.11-1.13.17.jar:1.13.17]
	at org.apache.flink.client.deployment.application.ApplicationDispatcherBootstrap.lambda$runApplicationAsync$1(ApplicationDispatcherBootstrap.java:219) ~[flink-dist_2.11-1.13.17.jar:1.13.17]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_232]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_232]
	at org.apache.flink.runtime.concurrent.akka.ActorSystemScheduledExecutorAdapter$ScheduledFutureTask.run(ActorSystemScheduledExecutorAdapter.java:159) [flink-dist_2.11-1.13.17.jar:1.13.17]
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:40) [flink-dist_2.11-1.13.17.jar:1.13.17]
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:44) [flink-dist_2.11-1.13.17.jar:1.13.17]
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260) [flink-dist_2.11-1.13.17.jar:1.13.17]
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339) [flink-dist_2.11-1.13.17.jar:1.13.17]
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979) [flink-dist_2.11-1.13.17.jar:1.13.17]
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107) [flink-dist_2.11-1.13.17.jar:1.13.17]
Caused by: org.apache.flink.client.deployment.application.ApplicationExecutionException: The application contains no execute() calls.
	at org.apache.flink.client.deployment.application.ApplicationDispatcherBootstrap.runApplicationEntryPoint(ApplicationDispatcherBootstrap.java:257) ~[flink-dist_2.11-1.13.17.jar:1.13.17]
	... 10 more
2022-09-06 11:59:46,427 INFO  org.apache.flink.runtime.entrypoint.ClusterEntrypoint        [] - Shutting YarnApplicationClusterEntryPoint down with application status FAILED. Diagnostics null.
2022-09-06 11:59:46,429 INFO  org.apache.flink.runtime.jobmaster.MiniDispatcherRestEndpoint [] - Shutting down rest endpoint.
```

Reply:
How was the job submitted with such error stack trace ?

Reply:
Application is submitted and executed in a Flink Job Cluster.

The command that is used to start the job:
```shell
/bin/flink run-application \
	-t yarn-application \
	-p 4 \
	-c org.apache.hudi.sink.compact.HoodieFlinkCompactor \
	hdfs://hudi-flink.jar \
	--path hdfs://hudi_mor_compaction_test/ \
	--compaction-tasks 4
```

Reply:
Let's remove the dummy execution logic first to avoid confusion (a successful execution that does nothing), an exception throws here does not affect anything right ? And the user can know that there is no compaction plan that needs to execute.

Reply:
While the exception throw does not affect anything (For most parts). 

I would argue that it is more confusing for the it to throw an exception when there are no compaction/clustering plans to execute.

If downstream applications are used to determine the status of the job, we would deem that this job has failed given that an exception is thrown (unless we explicitly scan the log to look for keywords to indicate that the failure is intended). 

My goal here is to remove this such unintended errors to avoid operational confusion. In such a case, instead of adding an execution logic to avoid confusion, can we add an additional catch block to catch `ApplicationExecutionException`?

Reply:
> can we add an additional catch block to catch ApplicationExecutionException

We can if we can analyze the `no execute` error message.

Reply:
Updated the code for both `compactor` and `clusteringjob` to include a catch block like this:

```java
try {
  compactionScheduleService.compact();
} catch (ApplicationExecutionException aee) {
  if (aee.getMessage().contains(NO_EXECUTE_CALL_KEYWORD)) {
    LOG.info("Compaction is not performed");
  } else {
    throw aee;
  }
} catch (Exception e) {
  LOG.error("Got error running delta sync once. Shutting down", e);
  throw e;
} finally {
  LOG.info("Shut down hoodie flink compactor");
}
```